### PR TITLE
fix(devtools-plugin): send action even if error was thrown

### PR DESCRIPTION
**Pull request by arturovt**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
Issue Number: #329

prev-devtools

## What is the new behavior?
after-devtools

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
@splincode @markwhitfeld 

This is very small bugfix for the old #329 issue. In the current behavior our `devtools-plugin` notifies Chrome extension only if everything is OK and no error is thrown. This PR fixes this behavior and notifies Chrome extension to log the state even if error was thrown. Given the following code:
```ts
catchError(error => {
  ctx.patchState({ counter });
  throw error;
})
```

This change will not be shown in the Chrome extension in the current behavior.

### CODE REVIEW

File: packages/devtools-plugin/src/devtools.plugin.ts [hunk 1]
```diff
@@ -57,6 +57,11 @@ export class NgxsReduxDevtoolsPlugin implements NgxsPlugin {
 
           this.devtoolsExtension!.send({ ...action, type }, newState);
         }
+      }),
+      catchError(error => {
+        const type = getActionTypeFromInstance(action);
+        this.devtoolsExtension!.send({ ...action, type }, this.store.snapshot());
+        throw error;
```
Comment (markwhitfeld) [inline packages/devtools-plugin/src/devtools.plugin.ts]: We could also get here if something in the `tap` part throws an error (like an error calling the dev tool).
Maybe we could use the `finally` pipe instead and combine the code?
Also, if there is an error for the `@@INIT` action would the dev tools extension not get initialized?

Comment (arturovt) [reply packages/devtools-plugin/src/devtools.plugin.ts]: The problem is that Rx doesn't pass anything as arguments to the `finalize` operator, no error and no `nextState`

Comment (markwhitfeld) [reply packages/devtools-plugin/src/devtools.plugin.ts]: Hmmm. Then could you catch before the `tap`?

Comment (arturovt) [reply packages/devtools-plugin/src/devtools.plugin.ts]: I'll try

File: packages/devtools-plugin/src/devtools.plugin.ts [hunk 2]
```diff
@@ -47,6 +47,11 @@ export class NgxsReduxDevtoolsPlugin implements NgxsPlugin {
     }
 
     return next(state, action).pipe(
+      catchError(error => {
+        const type = getActionTypeFromInstance(action);
+        this.devtoolsExtension!.send({ ...action, type }, this.store.snapshot());
+        throw error;
+      }),
```
Comment (markwhitfeld) [inline packages/devtools-plugin/src/devtools.plugin.ts]: Ok, this is better but what if the action that errors here was the `@@INIT` action?
The dev tool would still need to init. Should we maybe move the init code somewhere else?

Comment (markwhitfeld) [reply packages/devtools-plugin/src/devtools.plugin.ts]: I refactored so that the same logic is shared when sending to the dev tools.

**Comment by markwhitfeld** [comment]
@arturovt @splincode Please review my changes. Just made the logic for sending to the dev tool shared.

**Comment by arturovt** [comment]
@markwhitfeld I'm OK with your changes :)

**Pull request merged**